cla_backend/apps/reports/tests/test_reports.py

```diff
@@ -76,18 +76,64 @@ def test_mi_survey_dom1_extract_organisation_column(self):
 
 
 class ReportsDateRangeValidationWorks(TestCase):
-    def test_range_validation_works(self):
+    def test_valid_date_range(self):
+        class T(reports.forms.DateRangeReportForm):
+            max_date_range = 5
+
+        now = datetime.datetime.now()
+        i = T(data={"date_from": now - datetime.timedelta(days=4), "date_to": now})
+        self.assertTrue(i.is_valid())
+        self.assertEqual(i.errors.keys(), [])
+
+    def test_invalid_date_range(self):
         class T(reports.forms.DateRangeReportForm):
             max_date_range = 5
 
         now = datetime.datetime.now()
         i = T(data={"date_from": now - datetime.timedelta(days=5), "date_to": now})
         self.assertFalse(i.is_valid())
-        self.assertEqual(
-            i.errors, {u"__all__": [u"The date range (6 days, 0:00:00) should span no more than 5 working days"]}
+        self.assertEqual(i.errors.keys(), ["__all__"])
+        self.assertEqual(len(i.errors["__all__"]), 1)
+        self.assertIn(
+            i.errors[u"__all__"][0],
+            [
+                u"The date range (6 days, 0:00:00) should span no more than 5 working days",
+                u"The date range (6 days, 1:00:00) should span no more than 5 working days",
+                u"The date range (5 days, 23:00:00) should span no more than 5 working days",
+            ],
         )
-        i2 = T(data={"date_from": now - datetime.timedelta(days=1), "date_to": now})
-        self.assertTrue(i2.is_valid())
+
+    def test_valid_date_range_clocks_going_forward(self):
+        class T(reports.forms.DateRangeReportForm):
+            max_date_range = 5
+
+        start = datetime.datetime(2019, 3, 29, 12)
+        i = T(data={"date_from": start, "date_to": start + datetime.timedelta(days=4)})
+        self.assertTrue(i.is_valid())
+
+    def test_valid_date_range_clocks_going_back(self):
+        class T(reports.forms.DateRangeReportForm):
+            max_date_range = 5
+
+        start = datetime.datetime(2019, 10, 25, 12)
+        i = T(data={"date_from": start, "date_to": start + datetime.timedelta(days=4)})
+        self.assertTrue(i.is_valid())
+
+    def test_invalid_date_range_clocks_going_forward(self):
+        class T(reports.forms.DateRangeReportForm):
+            max_date_range = 5
+
+        start = datetime.datetime(2019, 3, 29, 12)
+        i = T(data={"date_from": start, "date_to": start + datetime.timedelta(days=5)})
+        self.assertFalse(i.is_valid())
+
+    def test_invalid_date_range_clocks_going_back(self):
+        class T(reports.forms.DateRangeReportForm):
+            max_date_range = 5
+
+        start = datetime.datetime(2019, 10, 25, 12)
+        i = T(data={"date_from": start, "date_to": start + datetime.timedelta(days=5)})
```

Comment:
I read this test as validating `max_date_range = 5` in relation to `2019, 10, 25, 12 + timedelta(days=5)` is invalid.

I'm afraid I don't understand what this means. Why is this invalid?

Reply:
(What probably confuses me is that we're calculating the `date_to` with `.timedelta` and I don't understand why.)

Reply:
It's cos the date range is inclusive, so today to tomorrow is a 2-day range, today to (today + 2) is a 3-day range, and so on.

Reply:
> (What probably confuses me is that we're calculating the `date_to` with `.timedelta` and I don't understand why.)

Just a hangover from the tests which use `now()` as the start, rather than a fixed date

Reply:
What do you think about changing the test to use fixed dates?
`"date_from": 2019-10-25 12:00 BST, "date_to": 2019-10-30 12:00 GMT` would make it easier for me to understand what's being asserted here.

Reply:
Updated